### PR TITLE
Feature/snippets/dynamism/enable mha token f32

### PR DESCRIPTION
**Pull request by IvanNovoselov**
### Details:
 - *Supplementary PR to facilitate the review of [PR#25500](https://github.com/openvinotoolkit/openvino/pull/25500)*

### Tickets:
 - *-*

### CODE REVIEW

File: src/common/snippets/src/utils/utils.cpp [hunk 1]
```diff
@@ -103,10 +103,10 @@ auto get_non_scalar_constant_count_for_fq(const std::shared_ptr<ov::op::v0::Fake
...
 
 bool broadcast_merge_dim(size_t& dst, const size_t& d1, const size_t& d2) {
-    if (d1 == d2 || d1 == 1 || is_dynamic_value(d2)) {
+    if (d1 == d2 || d1 == 1 || (is_dynamic_value(d1) && d2 != 1)) {
         dst = d2;
         return true;
-    } else if (d2 == 1 || is_dynamic_value(d1)) {
+    } else if (d2 == 1 || is_dynamic_value(d2)) {
         dst = d1;
         return true;
```
Comment (IvanNovoselov) [inline src/common/snippets/src/utils/utils.cpp]: This kind of changes suggest that we need shape infer tests in snippets to check dynamism handling, and probably other corner cases. What do you think?

File: src/common/snippets/src/lowered/pass/propagate_subtensors.cpp [hunk 2]
```diff
@@ -175,13 +175,13 @@ bool UpdateSubtensors::run(LinearIR& linear_ir, LinearIR::constExprIt begin, Lin
...
 std::shared_ptr<pass::PassBase> UpdateSubtensors::merge(const std::shared_ptr<pass::PassBase>& other) {
-    const auto merged_pass = std::make_shared<UpdateSubtensors>(m_tail_size);
-    if (other == nullptr)
-        return merged_pass;
+    if (!other)
+        return std::make_shared<UpdateSubtensors>(*this);
     const auto casted_pass = ov::as_type_ptr<UpdateSubtensors>(other);
-    if (!casted_pass || m_tail_size != casted_pass->m_tail_size)
+    size_t broadcasted_size;
+    if (!casted_pass || !ov::snippets::utils::broadcast_merge_dim(broadcasted_size, m_tail_size, casted_pass->m_tail_size))
```
Comment (IvanNovoselov) [inline src/common/snippets/src/lowered/pass/propagate_subtensors.cpp]: It's not obvious that merging of passes with `tails_size = 1` and `tail_size = 20` should yield `tail_size = 20`.
Apparently these changes are mainly for dynamic pipeline? If so, should we check that one of the tail sizes is dynamic?
In any case, since the merging routine is non-trivial, we should explain it in details in the pass in-code docs.

Comment (IvanNovoselov) [reply src/common/snippets/src/lowered/pass/propagate_subtensors.cpp]: The same questions are valid for `UpdateMemoryAccessCounts` and `SetFillOffset`.

File: src/common/snippets/src/lowered/pass/propagate_subtensors.cpp [hunk 3]
```diff
@@ -175,13 +175,13 @@ bool UpdateSubtensors::run(LinearIR& linear_ir, LinearIR::constExprIt begin, Lin
 }
 
 std::shared_ptr<pass::PassBase> UpdateSubtensors::merge(const std::shared_ptr<pass::PassBase>& other) {
-    const auto merged_pass = std::make_shared<UpdateSubtensors>(m_tail_size);
-    if (other == nullptr)
-        return merged_pass;
+    if (!other)
+        return std::make_shared<UpdateSubtensors>(*this);
```
Comment (IvanNovoselov) [inline src/common/snippets/src/lowered/pass/propagate_subtensors.cpp]: Why do we need to create a new pass instance for in such cases? 
Can't we just return `shared_from_this()`?
Of course, we need to derive `PassBase` from `std::enable_shared_from_this<PassBase>` to be able to do so, but it's a perfectly valid functionality.

File: src/common/snippets/src/lowered/pass/iter_handler.cpp
Comment (IvanNovoselov) [inline src/common/snippets/src/lowered/pass/iter_handler.cpp:1]: see propagate_subtensor relevant comments.

File: src/common/snippets/tests/include/utils/broadcast_dim_merge.hpp
Comment (IvanNovoselov) [inline src/common/snippets/tests/include/utils/broadcast_dim_merge.hpp:1]: Thanks for adding the `broadcast_merge_dim` test, this is a step in the right direction, since `ov::snippets::utils::broadcast_merge_dim` is used in many other passes.
However, since it's just a utility function, there is no guarantee that it would actually be used to implement shape  inference or subtensor propagation (or, you name it) passes. 
Therefore, do you think we can add a shape infer unit test as well? Or probably other critical pass unit tests (e.g. to reproduce introduced `merge` behavior)?

File: src/plugins/intel_cpu/src/node.h
Comment (IvanNovoselov) [inline src/plugins/intel_cpu/src/node.h:1]: I think we formally need someone else's approve from CPU team for this change. Can you ask Maxim or Egor to take a look please?

File: src/plugins/intel_cpu/src/nodes/subgraph.cpp [hunk 4]
```diff
@@ -761,6 +761,7 @@ void Subgraph::prepareParams() {
 
     auto builder = [this, cache](const SubgraphKey& key) -> std::shared_ptr<SubgraphExecutor> {
         const auto& snippet = subgraph_attrs->snippet;
+        const auto scratchpad = getScratchPad();
```
Comment (IvanNovoselov) [inline src/plugins/intel_cpu/src/nodes/subgraph.cpp:755]: Since we're refreshing the pointer every time the builder is called, there's no need to copy the whole shared ptr by value, we can use const reference instead.

File: src/plugins/intel_cpu/src/nodes/subgraph.cpp [hunk 5]
```diff
@@ -196,8 +196,8 @@ class SubgraphDynamicSpecializedExecutor : public Subgraph::SubgraphExecutor {
         }
     }
 
-    void init_runtime_params(const std::shared_ptr<CPURuntimeConfig>& snippet_config) override {
-        SubgraphExecutor::init_runtime_params(snippet_config);
+    void init_runtime_params(const std::shared_ptr<CPURuntimeConfig>& snippet_config,  const DnnlScratchPadPtr& scratchpad) override {
+        SubgraphExecutor::init_runtime_params(snippet_config, scratchpad);
```
Comment (IvanNovoselov) [inline src/plugins/intel_cpu/src/nodes/subgraph.cpp]: As we discussed previously, the `init_runtime_params` is not really needed: we can (and should) do all this in `SubgraphExecutor` and `SubgraphDynamicSpecializedExecutor` constructors.
Do you think we can eliminate the redundant method now?

File: src/plugins/intel_cpu/src/nodes/subgraph.cpp [hunk 6]
```diff
@@ -97,17 +98,16 @@ class SubgraphStaticExecutor : public Subgraph::SubgraphExecutor {
...
 
         for (size_t i = 0; i < dstMemPtrs.size(); i++)
             call_args.dst_ptrs[i] = dstMemPtrs[i]->getDataAs<uint8_t>() + m_start_offset_out[i];
 
-        if (m_buffer_scratchpad_size > 0) {
-            call_args.buffer_scratchpad_ptr =
-                    reinterpret_cast<uint8_t*>(m_buffer_scratchpad.data()) + parallel_get_thread_num() * m_buffer_scratchpad_size;
-        }
+        if (m_buffer_scratchpad_size > 0)
+            call_args.buffer_scratchpad_ptr = m_buffer_scratchpad->getDataAs<uint8_t>() + ithr * m_buffer_scratchpad_size;
```
Comment (IvanNovoselov) [inline src/plugins/intel_cpu/src/nodes/subgraph.cpp]: Do you think it's worth to create a corresponding inline function in the base class to avoid duplication?
Like `inline void update_scratchpad_ptr(jit_snippets_call_args& call_args, size_t ithr)` or even better:
`inline void update_scratchpad_ptr(void*& scratchpad_ptr, size_t ithr)`

File: src/common/snippets/src/lowered/pass/propagate_subtensors.cpp (hunk 3 above)
Comment (a-sidorova) [reply src/common/snippets/src/lowered/pass/propagate_subtensors.cpp]: > Why do we need to create a new pass instance for in such cases?
Can't we just return shared_from_this()?

I had the same question! Because I just moved the previous logic and I wanted to ask the author of these handlers, Vladislav. But he is OOO. So I decided to leave it so. Let's try to do that 😃 Thanks

File: src/common/snippets/src/lowered/pass/propagate_subtensors.cpp (hunk 2 above)
Comment (a-sidorova) [reply src/common/snippets/src/lowered/pass/propagate_subtensors.cpp]: Great catch! I forgot about scalar value! For sure, it's valid only for dynamic case.
Created the util `merge_dynamic_dim` that doesn't allow broadcasting for static dimensions
Thanks!

File: src/common/snippets/src/utils/utils.cpp (hunk 1 above)
Comment (a-sidorova) [reply src/common/snippets/src/utils/utils.cpp]: Exactly! In this PR I created the unit test `BroadcastMergeDimTest` at least. Just to validate this method ( I fixed this method two times). As for shape inference tests, I left the comment below.

File: src/common/snippets/tests/include/utils/broadcast_dim_merge.hpp
Comment (a-sidorova) [reply src/common/snippets/tests/include/utils/broadcast_dim_merge.hpp:1]: Totally agree! Since `ov::snippets::utils::broadcast_merge_dim` is used in many places, I decided to create the simplest unit test for this utility as first step. But seems like we should have the unit tests for shape inference.
I created the ticket: https://jira.devtools.intel.com/browse/CVS-150013
Probably when we support new op with new shape inference rules, we can come back to this ticket 🤔

File: src/plugins/intel_cpu/src/node.h
Comment (a-sidorova) [reply src/plugins/intel_cpu/src/node.h:1]: I wanted to ask someone when this PR will be rebased on the merged `SplitLoops` to the master and ready for review. Thanks for the reminder

File: src/plugins/intel_cpu/src/nodes/subgraph.cpp (hunk 4 above)
Comment (a-sidorova) [reply src/plugins/intel_cpu/src/nodes/subgraph.cpp:755]: But `Node::getScratchPad` returns the copy of shared pointer as `GraphContext::getScratchPad` does it.
Since `Node::getScratchPad` returns the copy, I use `const auto scratchpad` here 🤔 

If we want to use `const ref` here, the corresponding methods should return this type.
I decided to not change it since it's CPU part. However, we can ask Egor or Maxim in review.

Example: https://coliru.stacked-crooked.com/a/7334b45304f10d47

File: src/plugins/intel_cpu/src/nodes/subgraph.cpp (hunk 5 above)
Comment (a-sidorova) [reply src/plugins/intel_cpu/src/nodes/subgraph.cpp]: Agree, removed. Thanks!

File: src/plugins/intel_cpu/src/nodes/subgraph.cpp (hunk 6 above)
Comment (a-sidorova) [reply src/plugins/intel_cpu/src/nodes/subgraph.cpp]: Yeah, good idea. Done, thanks

**Review by a-sidorova — COMMENTED** [review]
@IvanNovoselov thanks a lot for the review!
May I ask you please to take a look again?

File: src/common/snippets/include/snippets/utils/utils.hpp [hunk 7]
```diff
@@ -124,6 +124,11 @@ std::string vector2str(const std::vector<T>& values) {
 
 bool broadcast_merge_dim(size_t& dst, const size_t& d1, const size_t& d2);
 
+// If one of dim is dynamic at least, return the another dim (it might be also dynamic or static)
+// If the both dims are static, they must be equal - this's the difference from the utility above
+// Can be used in SpecificLoopIterationHandlers
```
Comment (IvanNovoselov) [inline src/common/snippets/include/snippets/utils/utils.hpp:129]: Minor:
```suggestion
// If one of the dims is dynamic, return the other dim (might also be dynamic)
// If both dims are static, they must be equal - this is the difference from the utility above
// Can be used in SpecificLoopIterationHandlers
```

File: src/common/snippets/src/lowered/pass/iter_handler.cpp [hunk 8]
```diff
@@ -49,13 +49,13 @@ bool UpdateMemoryAccessCounts::run(LinearIR& linear_ir, LinearIR::constExprIt be
 }
 
 std::shared_ptr<pass::PassBase> UpdateMemoryAccessCounts::merge(const std::shared_ptr<pass::PassBase>& other) {
-    const auto merged_pass = std::make_shared<UpdateMemoryAccessCounts>(m_count);
-    if (other == nullptr)
-        return merged_pass;
+    if (!other)
+        return this->shared_from_this();
```
Comment (IvanNovoselov) [inline src/common/snippets/src/lowered/pass/iter_handler.cpp:53]: Why would one write `this` here?
I mean, it's not wrong, but why? It's extra effort both to write and to read 😁 
It's assumed implicitly, like for all other member functions

File: src/common/snippets/src/lowered/pass/iter_handler.cpp [hunk 9]
```diff
@@ -49,13 +49,13 @@ bool UpdateMemoryAccessCounts::run(LinearIR& linear_ir, LinearIR::constExprIt be
...
     const auto casted_pass = ov::as_type_ptr<UpdateMemoryAccessCounts>(other);
-    if (!casted_pass || m_count != casted_pass->m_count)
+    size_t merged_count;
+    if (!casted_pass || !ov::snippets::utils::merge_dynamic_dim(merged_count, m_count, casted_pass->m_count))
```
Comment (IvanNovoselov) [inline src/common/snippets/src/lowered/pass/iter_handler.cpp:56]: It looks like we don't need to actually cast the pass - we never use the casted pointer.
Can you replace it with `ov::is_type<...>` please? It would work faster. After all, this is one of the reasons why we have custom rt info in the project.

We can also show off a little, and write a generic check that would work for all types:
```
if (!ov::is_type<std::remove_pointer<decltype(this)>::type>(other) || ...)
```

Valid for all affected handlers: this file and propagate_subtensors.

File: src/plugins/intel_cpu/src/nodes/subgraph.h [hunk 10]
```diff
@@ -122,18 +122,23 @@ class Subgraph::SubgraphExecutor {
...
 protected:
-    void parallel_for6d(const std::function<void(jit_snippets_call_args&)>& initializer,
+    void parallel_for6d(const std::function<void(jit_snippets_call_args&, size_t)>& initializer,
                         const std::function<void(jit_snippets_call_args&, const size_t*)>& caller);
-    void parallel_forNd(const std::function<void(jit_snippets_call_args&)>& initializer,
+    void parallel_forNd(const std::function<void(jit_snippets_call_args&, size_t)>& initializer,
                         const std::function<void(jit_snippets_call_args&, const size_t*)>& caller);
 
-    virtual void init_runtime_params(const std::shared_ptr<CPURuntimeConfig>& snippet_config);
+    inline void update_scratchpad_ptr(void*& scratchpad_ptr, size_t ithr) {
```
Comment (IvanNovoselov) [inline src/plugins/intel_cpu/src/nodes/subgraph.h:138]: It looks like we can make it `const`. Can't we?

File: src/common/snippets/tests/include/utils/broadcast_dim_merge.hpp
Comment (IvanNovoselov) [reply src/common/snippets/tests/include/utils/broadcast_dim_merge.hpp:1]: I'm not sure we need to wait for a new op 🙂
Ok, I don't insist on doing it in this PR. Especially considering our plans with respect to symbolic shape inference.

File: src/plugins/intel_cpu/src/nodes/subgraph.cpp (hunk 4 above)
Comment (IvanNovoselov) [reply src/plugins/intel_cpu/src/nodes/subgraph.cpp:755]: > If we want to use const ref here, the corresponding methods should return this type.

I've never heard about this rule. Why?

In my opinion this particular case is more about the intent declaration, rather than an actual extra copy elimination (since any decent compiler will produce optimal code here due to copy elision optimizations). When you copy an object by reference, you you usually need either to request some info from the object, or pass it further to other classes/functions (our case). If you (const) copy an object by value you assume that the object will change soon, and you need to save a particular state of the object before it does change.

In general, I follow Meyers advice (see [this](https://stackoverflow.com/a/270435) thread for a discussion): try to use references for non-build-in types where possible. However I agree that in this case this is not that important.

File: src/common/snippets/include/snippets/utils/utils.hpp (hunk 7 above)
Comment (a-sidorova) [reply src/common/snippets/include/snippets/utils/utils.hpp:129]: Corrected, thanks

File: src/common/snippets/src/lowered/pass/iter_handler.cpp (hunk 8 above)
Comment (a-sidorova) [reply src/common/snippets/src/lowered/pass/iter_handler.cpp:53]: Removed, thanks

File: src/common/snippets/src/lowered/pass/iter_handler.cpp (hunk 9 above)
Comment (a-sidorova) [reply src/common/snippets/src/lowered/pass/iter_handler.cpp:56]: > It looks like we don't need to actually cast the pass - we never use the casted pointer.

No, we have to cast the pass pointer to the specific type to get attributes (for example, `m_count` here, `m_offset` in `SetFillOffset` and `m_tail_size` in `UpdateSubtensors`)

File: src/plugins/intel_cpu/src/nodes/subgraph.h (hunk 10 above)
Comment (a-sidorova) [reply src/plugins/intel_cpu/src/nodes/subgraph.h:138]: Sure, added. Thanks

File: src/plugins/intel_cpu/src/nodes/subgraph.cpp (hunk 4 above)
Comment (a-sidorova) [reply src/plugins/intel_cpu/src/nodes/subgraph.cpp:755]: >try to use references for non-build-in types where possible

I fully agree with `use const ref non-build-in types where possible`. But in this case (when the function return a copy), it looks redundant (in my opinion) 🤔

Moreover, a few lines above there is `const auto cache = context->getParamsCache();` <- this is the same behavior 😄

**Comment by a-sidorova** [comment]
I rebased the original PR and added the changes to there https://github.com/openvinotoolkit/openvino/pull/25500
